python/neuroglancer/ingest/super_downsample.py

```diff
@@ -0,0 +1,111 @@
+from neuroglancer import chunks, downsample, downsample_scales
+
+from base import Storage
+
+import lib
+from lib import clamp, mkdir, list_shape, xyzrange, Vec3, Bbox, map2, min2, max2
+from lib import GCLOUD_PROJECT_NAME, COMMON_STAGING_DIR
+
+from PIL import Image
+
+def generate_chunks(img, chunk_sizes):
+  chunk_sizes = Vec3(*chunk_sizes)
+  volume_size = Vec3(*img.shape)
+
+  for startpt in xyzrange( (0,0,0), volume_size, chunk_sizes ):
+    endpt = min2(startpt + chunk_sizes, volume_size)
+
+    chunkimg = img[ startpt.x:endpt.x, startpt.y:endpt.y, startpt.z:endpt.z ]
+
+    endpt = min2(endpt, volume_size - 1)
+
+    filename = '{}-{}_{}-{}_{}-{}'.format(
+      startpt.x, endpt.x,
+      startpt.y, endpt.y,
+      startpt.z, endpt.z
+    ) 
+
+    yield chunkimg.T, filename # transpose chunkimg to make it in fortran order
+
+def generate_downsamples(dataset_name, layer, starting_mip=-1):
+  vol = GCloudVolume(dataset_name, layer, mip=starting_mip)
+
+  scales = downsample_scales.compute_near_isotropic_downsampling_scales(
+    size=vol.mip_shape(starting_mip),
+    voxel_size=vol.mip_resolution(starting_mip),
+    dimensions_to_downsample=[0,1,2],
+    max_downsampling=int(reduce(operator.mul, vol.shape / vol.underlying )),
+  )
+
+  scales = scales[1:] # omit (1,1,1)
+
+  base_scale_factor = vol.mip_resolution(starting_mip) / vol.mip_resolution(0)
+
+  compress = (vol.layer_type == 'segmentation')
+
+  storage = Storage(dataset_name, layer, compress)
+
+  renderbuffer = vol[:,:,:]
+
+  for factor3 in scales:
+    scale = vol.addScale(factor3 * base_scale_factor)
+    downsampled_img = downsample.downsample_with_striding(renderbuffer, factor3)
+
+    image_chunks = generate_chunks(downsampled_img, vol.underlying)
+
+    for img_chunk, filename in tqdm(image_chunks, desc="{} Chunks".format(factor3)):
+      if scale["encoding"] == "jpeg":
+        encoded = chunks.encode_jpeg(img_chunk)
+      elif scale["encoding"] == "npz":
+        encoded = chunks.encode_npz(img_chunk)
+      elif scale["encoding"] == "raw":
+        encoded = chunks.encode_raw(img_chunk)
+      else:
+        raise NotImplemented
+
+      storage.add_file(filename, encoded)
+
+    storage.flush(scale['key'])
+
+    vol.commit()
+
+def generate_jpegs(dataset_name, layer):
+  """used for debugging"""
+  vol = GCloudVolume(dataset_name, layer, mip=-1)
+
+  zslice = slice(0, 50, 1)
+
+  renderbuffer = vol[:,:,zslice]
+
+  for z in tqdm(xrange(zslice.stop)):
+    img = renderbuffer[:,:,z]
+    # shape = img.shape
+    # reshaped = img.reshape(shape[0] * shape[1], 1)
+
+    # discovered that downloaded cube is in a weird rotated state.
+    # it requires a 90deg counterclockwise rotation on xy plane (leaving z alone)
+    # followed by a flip on Y
+    img = np.flipud(np.rot90(img, 1)) 
+    img = Image.fromarray(img)
+    print 'saving as: ', './staging/jpegs/{}.jpeg'.format(z)
+    img.save('./staging/jpegs/{}.jpeg'.format(z), "JPEG")
+
+if __name__ == "__main__":
```

Comment:
On the contrary, they're what run the program. ;) You uncomment and adjust them to your liking (currently). I can turn this into a real script too with command line arguments.